Merging events domain and name (#473)
open-telemetry/semantic-conventions: docs/general/events.md

## General event attributes

Events are recorded as LogRecords that are shaped
in a special way: Event LogRecords have the attribute

> `event.name` consisting of a `namespace` and `name` that

Reviewer: ```suggestion
`event.name` consisting of a dot-separated `namespace` and `name` that
```

> uniquely identifies the event (and possibly other LogRecord attributes).

Reviewer 2: ```suggestion
uniquely identifies the event. Events may also have other LogRecord attributes.
```

The `namespace` portion is used to logically separate events from different
systems. For example, to record Events from browser apps, mobile apps and
Kubernetes, we could use `browser`, `device` and `k8s` as the `namespace` for their
Events. This provides a clean separation of semantics for events in each of the
domains.

Within a particular `namespace`, the `name` portion identifies the event.

> Events with same `namespace` and `name` are structurally similar to one another. For
> example, some namespaces could have well-defined schema for their events based on
> event names.

Reviewer 2: ```suggestion
Events with same `event.name` are structurally similar to one another.
```

## General event attributes

Events are recorded as LogRecords that are shaped
in a special way: Event LogRecords have the attribute
`event.name` consisting of a dot-separated `namespace` and `name` that
uniquely identifies the event. Events may also have other LogRecord attributes.

> The `namespace` portion is used to logically separate events from different

Reviewer 3: Rather than define the namespace concept in this document, I think we should link out to the [attribute naming](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/common/attribute-naming.md) document and say something to the effect of: "Event names are subject to the same rules as [attribute names](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/common/attribute-naming.md). Notably, event names are namespaced to avoid collisions."

Author: @jack-berg take another look pls. I also updated the general info to remove the blurbs around defining namespace concept.

Reviewer 3: 👍

<!-- semconv event -->
| Attribute  | Type | Description  | Examples  | Requirement Level |
|---|---|---|---|---|

> | `event.name` | string | Uniquely identifies the event. | `browser.mouse.click`; `browser.exception` | Required |

Reviewer 3: Can we find a different example than `browser.exception`? I don't envision having different exception events for different runtimes. How about using the `device.app.lifecycle` event name defined here? https://github.com/open-telemetry/semantic-conventions/blob/main/docs/mobile/events.md

(Side note, we'll want to update the device app.lifecycle event and any others like it that specify a event.domain.

Reviewer 4: +1

Reviewer 4: How about `device.app.lifecycle` or `mobile.app.lifecycle` ?

Author: I think we need to decide on what our prefixes should look like before I change this. Do we need to use `browser` and `mobile` or `device`? What other prefixes would be acceptable and how do we tell consumers about them?

Reviewer 3: This PR is all about the merging of `event.domain` and `event.name`. The natural way to merge these is to concatenate `event.domain + event.name`. So let's update the [mobile lifecycle events](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/mobile/events.md#lifecycle-instrumentation) to have event.names of `device.app.lifecycle`.

> What other prefixes would be acceptable and how do we tell consumers about them?

The [attribute naming](https://github.com/open-telemetry/opentelemetry-specification/tree/v1.26.0/specification/common/attribute-naming.md) document you've linked to defines the rules for `event.name`.

Author: Should we do that in this PR or a separate PR?

Reviewer 3: I think this PR. The [mobile lifecylce events](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/mobile/events.md#lifecycle-instrumentation) is the only place where an `event.domain` / `event.name` is specified, so should be small in scope.

Author: Done.

Reviewer 3: Marking this comment as "unresolved" because `browser.exception` is still used as an example. Can we switch to `device.app.lifecycle` instead?

Author: Should be good now.

> | `event.name` | string | Uniquely identifies the event. | `browser.mouse.click`; `browser.exception` | Required |

Reviewer 3: ```suggestion
| `event.name` | string | Identifies the class / type of event. | `browser.mouse.click`; `browser.exception` | Required |
```
The phrasing of "Uniquely identifies" reads like this could be some sort of UUID type identifier which is unique for every distinct record.

> Event names are subject to the same rules as [attribute names](https://github.com/open-telemetry/opentelemetry-specification/tree/v1.26.0/specification/common/attribute-naming.md).

Author: Should we not point to a specific version of the specification?

> Event names are subject to the same rules as [attribute names](https://github.com/open-telemetry/opentelemetry-specification/tree/v1.26.0/specification/common/attribute-naming.md).

Reviewer 2: No, I don't think we should. Spec is versioned all together, all of its documents. If the linked document is updated it should apply to this doc too.

> Event names are subject to the same rules as [attribute names](https://github.com/open-telemetry/opentelemetry-specification/tree/v1.26.0/specification/common/attribute-naming.md).

Author: The specs are linked to in a lot of other docs in this same manner. Should we update all of them and is that necessarily an ask or work that is outside the scope of this PR?

> Event names are subject to the same rules as [attribute names](https://github.com/open-telemetry/opentelemetry-specification/tree/v1.26.0/specification/common/attribute-naming.md).

Reviewer 2: No need to update other links in this PR.

I wonder if any reviewer has any arguments in favour of linking to a specific version? I thought it is a mistake but since it is in many places perhaps it is deliberate and in that case I wonder what the reason is.

> Event names are subject to the same rules as [attribute names](https://github.com/open-telemetry/opentelemetry-specification/tree/v1.26.0/specification/common/attribute-naming.md).

Reviewer 5: Linking to a specific version will be confusing unless the version is part of the link text and is specifically called out.

> Event names are subject to the same rules as [attribute names](https://github.com/open-telemetry/opentelemetry-specification/tree/v1.26.0/specification/common/attribute-naming.md).

Author: https://github.com/open-telemetry/semantic-conventions/issues/549 opened to update the other links in the repo.

<!-- semconv event -->
| Attribute  | Type | Description  | Examples  | Requirement Level |
|---|---|---|---|---|
| `event.name` | string | Identifies the class / type of event. [1] | `browser.mouse.click`; `device.app.lifecycle` | Required |

> **[1]:** Event names are subject to the same rules as [attribute names](https://github.com/open-telemetry/opentelemetry-specification/tree/v1.26.0/specification/common/attribute-naming.md). Notably, event names are namespaced to avoid collisions and provide a clean separation of semantics for events in separate domains like browser, mobile, and kubernetes.

Reviewer 2: I think the link needs to go to main branch.

Author: https://github.com/open-telemetry/semantic-conventions/issues/549 opened to update the other links in the repo.

Reviewer 2: I was wrong, see https://github.com/open-telemetry/semantic-conventions/pull/473#discussion_r1401926563

<!-- semconv event -->
| Attribute  | Type | Description  | Examples  | Requirement Level |
|---|---|---|---|---|
| `event.name` | string | Consisting of a `namespace` and `name`; uniquely identifies the event. | `browser.mouse.click`; `browser.exception` | Required |

> **[1]:** Events across different namespaces may have same name, yet be unrelated events.

Reviewer 2: ```suggestion
```

The `namespace` portion of the name has the following list of well-known values. If one of them applies, then the respective value MUST be used, otherwise a custom value MAY be used.

| Value  | Description |
|---|---|
| `browser` | Events from browser apps |
| `device` | Events from mobile apps |

> | `k8s` | Events from Kubernetes |

Reviewer 2: I think we should get rid of this and instead say that event namespaces are shared with all other semconv namespaces.

Author: I removed this table and blurb for now and I think the following section pointing to the attribute naming is enough for now. I do think we will probably have a table in here at some point for well known namespace prefixes once we come to an agreement in the WG.
